// unirand/lib/utils/string_utils/uid/index.js
"use strict";

var _ksuid = _interopRequireDefault(require("./ksuid"));

var _xid = _interopRequireDefault(require("./xid"));

var _betterguid = _interopRequireDefault(require("./betterguid"));

var _sonyflake = _interopRequireDefault(require("./sonyflake"));

var _snowflake = _interopRequireDefault(require("./snowflake"));

var _ulid = _interopRequireDefault(require("./ulid"));

var _uuid = _interopRequireDefault(require("./uuid"));

var _shortuuid = _interopRequireDefault(require("./shortuuid"));

var _sno = _interopRequireDefault(require("./sno"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/**
 * Random UID generators
 * Created by Alexey S. Kiselev
 */
var ksuid = new _ksuid["default"]();
var xid = new _xid["default"]();
var betterguid = new _betterguid["default"]();
var sonyflake = new _sonyflake["default"]();
var snowflake = new _snowflake["default"]();
var ulid = new _ulid["default"]();
var uuid = new _uuid["default"]();
var shortuuid = new _shortuuid["default"]();
var sno = new _sno["default"]();
module.exports = {
  ksuid: ksuid,
  xid: xid,
  betterguid: betterguid,
  sonyflake: sonyflake,
  snowflake: snowflake,
  ulid: ulid,
  uuid: uuid,
  shortuuid: shortuuid,
  sno: sno
};
