// unirand/lib/utils/string_utils/uid/snowflake.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _prngProxy = _interopRequireDefault(require("../../../prng/prngProxy"));

var _murmur = _interopRequireDefault(require("../../hash/murmur3"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var murmur3Hash = new _murmur["default"]();
var isRunningInNode = typeof process !== 'undefined' && process.versions != null && process.versions.node != null;
var MAX_RANDOM1_VALUE = 0x3FF;
var LEADING_ZEROS = ['', '0', '00', '000', '0000', '00000', '000000', '0000000', '00000000', '000000000', '0000000000'];
var EXPECTED_RIGHT_SIZE = 10;

var RandomSnowflakeGenerator = /*#__PURE__*/function () {
  function RandomSnowflakeGenerator() {
    _classCallCheck(this, RandomSnowflakeGenerator);

    this.need_random1 = true;
    this.sequence = 0;

    if (isRunningInNode) {
      var os = require('os');

      if (os.hostname().length > 0) {
        this.need_random1 = false;
        this.random1 = murmur3Hash.hash(os.hostname(), 0) & MAX_RANDOM1_VALUE;
      }
    }
  }

  _createClass(RandomSnowflakeGenerator, [{
    key: "generateRandom",
    value: function generateRandom() {
      this._generateRandom1IfNeeded(_prngProxy["default"].random());

      return this._getSnowflake();
    }
  }, {
    key: "generateNext",
    value: function generateNext() {
      this._generateRandom1IfNeeded(_prngProxy["default"].next());

      return this._getSnowflake();
    }
  }, {
    key: "_generateRandom1IfNeeded",
    value: function _generateRandom1IfNeeded(random) {
      if (this.need_random1) {
        this.random1 = Math.floor(random * MAX_RANDOM1_VALUE);
      }
    }
  }, {
    key: "_getSnowflake",
    value: function _getSnowflake() {
      var res = this._getTimestampData();

      var right = res.right;
      right |= this.random1 << 12;
      this.sequence = this.sequence + 1 & 0xFFF;
      right |= this.sequence;
      var rightAsString = right.toString();
      return res.left + LEADING_ZEROS[EXPECTED_RIGHT_SIZE - rightAsString.length] + rightAsString;
    }
  }, {
    key: "_getTimestampData",
    value: function _getTimestampData() {
      var left = 0;
      var right;
      var ts = Date.now();
      right = ts % 3;
      ts = Math.floor(ts / 3);
      right |= (ts % 256 & 0x7F) << 2;
      ts = Math.floor(ts / 256);
      right <<= 22;

      for (var i = 0; i < 4; i += 1) {
        left = ts % 256 << 8 * i;
        ts = Math.floor(ts / 256);
      }

      left &= 0x7FFFFFFF;
      return {
        left: left.toString(),
        right: right
      };
    }
  }]);

  return RandomSnowflakeGenerator;
}();

var _default = RandomSnowflakeGenerator;
exports["default"] = _default;
